internal/pkg/pluginengine/change_helper.go

```diff
@@ -59,49 +59,60 @@ func drifted(a, b map[string]interface{}) bool {
-			changes = append(changes, generateCreateAction(&tool, description))
-		} else {
-			// tool found in the state
-			if drifted(tool.Options, state.Options) {
-				// tool's config differs from State's, Update
-				description := fmt.Sprintf("Tool < %s > config drifted from the state, will be updated.", tool.Name)
-				changes = append(changes, generateUpdateAction(&tool, description))
-			} else {
-				// tool's config is the same as State's
+	// 3. generate changes for each tool
+	for _, batch := range tools {
+		for _, tool := range batch {
+			state := smgr.GetState(statemanager.StateKeyGenerateFunc(&tool))
 
-				// read resource status
-				resource, err := Read(&tool)
-				if err != nil {
-					return changes, err
-				}
-
-				if resource == nil {
-					// tool exists in the state, but resource doesn't exist, Create
-					description := fmt.Sprintf("Tool < %s > state found but it seems the tool isn't created, will be created.", tool.Name)
-					changes = append(changes, generateCreateAction(&tool, description))
-				} else if drifted(resource, state.Resource) {
-					// resource drifted from state, Update
-					description := fmt.Sprintf("Tool < %s > drifted from the state, will be updated.", tool.Name)
+			if state == nil {
+				// tool not in the state, create, no need to Read resource before Create
+				description := fmt.Sprintf("Tool < %s > found in config but doesn't exist in the state, will be created.", tool.Name)
+				changes = append(changes, generateCreateAction(&tool, description))
+			} else {
+				// tool found in the state
+				if drifted(tool.Options, state.Options) {
+					// tool's config differs from State's, Update
+					description := fmt.Sprintf("Tool < %s > config drifted from the state, will be updated.", tool.Name)
 					changes = append(changes, generateUpdateAction(&tool, description))
 				} else {
-					// resource is the same as the state, do nothing
-					log.Debugf("Tool < %s > is the same as the state, do nothing.", tool.Name)
+					// tool's config is the same as State's
+
+					// read resource status
+					resource, err := Read(&tool)
+					if err != nil {
+						return changes, err
+					}
+
+					if resource == nil {
+						// tool exists in the state, but resource doesn't exist, Create
+						description := fmt.Sprintf("Tool < %s > state found but it seems the tool isn't created, will be created.", tool.Name)
+						changes = append(changes, generateCreateAction(&tool, description))
+					} else if drifted(resource, state.Resource) {
+						// resource drifted from state, Update
+						description := fmt.Sprintf("Tool < %s > drifted from the state, will be updated.", tool.Name)
+						changes = append(changes, generateUpdateAction(&tool, description))
+					} else {
+						// resource is the same as the state, do nothing
+						log.Debugf("Tool < %s > is the same as the state, do nothing.", tool.Name)
```

Comment:
does nothing？

Reply:
I think this is fine. It doesn't sound official enough, but it means something like "the tool is the same as the state, so we do nothing here (so nothing is done here)".